lib/ansible/modules/cloud/amazon/aws_s3_sync.py

```diff
@@ -0,0 +1,643 @@
+          type: string
+          sample: '"fffcc0bc5495e11184419846975badc9"'
+        storage_class:
+          description: Storage class of the object
+          returned: always
+          type: string
+          sample: STANDARD
+'''
+
+from copy import copy
+from datetime import datetime
+from fnmatch import fnmatch
+import hashlib
+import mimetypes
+import os
+import re
+from stat import ST_MTIME, ST_SIZE
+import time
+import tempfile
+
+from ansible.module_utils.aws.core import AnsibleAWSModule
+from ansible.module_utils.aws.s3 import dict_to_s3_extra_args
+from ansible.module_utils.ec2 import AWSRetry
+from ansible.module_utils.common.dict_transformations import camel_dict_to_snake_dict
+from ansible.module_utils._text import to_bytes
+
+try:
+    from dateutil import tz
+    HAS_DATEUTIL = True
+except ImportError:
+    HAS_DATEUTIL = False
+
+try:
+    from hashlib import md5
+    HAS_MD5 = True
+except ImportError:
+    try:
+        from md5 import md5
+        HAS_MD5 = True
+    except ImportError:
+        HAS_MD5 = False
+
+try:
+    from botocore.exceptions import BotoCoreError, ClientError
+except ImportError:
+    pass    # Handled by AnsibleAWSModule
+
+
+class S3Syncer(object):
+    def __init__(self, module):
+        self.module = module
+        self.check_mode = module.check_mode
+        self.params = module.params
+        self.changed = False
+
+        params = self.params
+        self.file_args = module._module.load_file_common_arguments(params)
+        self.directory_args = copy(self.file_args)
+        self.directory_args['mode'] = params['directory_mode']
+
+        self.mode = params['direction']
+        self.bucket = params['bucket']
+        self.path = params['path']
+        self.prefix = params['prefix']
+        self.patterns = params['patterns']
+        self.excludes = params['excludes']
+        self.delete = params['delete']
+
+        self.mimetypes = mimetypes.MimeTypes()
+        if params['mime_override']:
+            self.mimetypes.encodings_map = params['mime_encodings_map']
+            self.mimetypes.types_map = ({}, params['mime_types_map'])
+        else:
+            self.mimetypes.encodings_map.update(params['mime_encodings_map'])
+            self.mimetypes.types_map[1].update(params['mime_types_map'])
+
+        try:
+            self.s3 = module.client('s3')
+        except (BotoCoreError, ClientError) as e:
+            module.fail_json_aws(e, msg="Couldn't connect to S3")
+
+        self.objects = []
+
+    def _calculate_etag(self, entry):
+        if '-' in entry['s3']['e_tag']:
+            # Multi-part ETag; a hash of the hashes of each part.
+            parts = int(entry['s3']['e_tag'][1:-1].split('-')[1])
+            digests = []
+            with open(entry['path'], 'rb') as f:
+                for part_num in range(1, parts + 1):
+                    try:
+                        key_head = self.s3.head_object(Bucket=self.bucket, Key=entry['key'], PartNumber=part_num)
```

Comment:
Just curious, for large files that are being synced, does this add a considerable amount of overhead for getting the ETag for each part?

Reply:
I haven't measured it; I'm sure it can be significant if you've enabled ETag calculation and there are a large number of multi-part files, but I don't think there's a reliable way to guess the part size. (Of course, generating an ETag for S3 objects is undocumented and unreliable in the first place...)